core/src/main/java/io/grpc/internal/ConcurrentRescheduler.java

```diff
@@ -0,0 +1,182 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Schedules, reschedules or cancels a runnable. As opposed to {@link Rescheduler}, {@code
+ * ConcurrentRescheduler} does not execute {@code runnable.run()} in a serialized executor.
+ * Lazy-reschedule is not used here because that will add logic complexity and currently no usecase
+ * benefits much from that.
+ */
+@ThreadSafe
+final class ConcurrentRescheduler {
+  private final Runnable runnable;
+  private final ScheduledExecutorService scheduler;
+  private final Object lock;
+  @GuardedBy("lock")
+  @Nullable
+  private final Precondition precondition;
+
+  // Represents the currently scheduled task.
+  // Set to null when the task is cancelled;
+  // also set to null when the task starts to run;
+  // set to a new instance when a new task is scheduled;
+  // also set to a new instance when the task is rescheduled.
+  // Whenever cancelled or rescheduled, the old value MUST be marked with cancelled = true.
+  @GuardedBy("lock")
+  @CheckForNull
+  private FutureRunnable wakeUp;
+
+  /**
+   * Constructor.
+   *
+   * @param precondition schedule the runnable only when precondition.met() is true. Does not apply
+   *                     the precondition for rescheduling an existing task. Does not check the
+   *                     precondition when running the task.
+   */
+  ConcurrentRescheduler(
+      Runnable runnable, ScheduledExecutorService scheduler, @Nullable Precondition precondition,
+      Object lock) {
+    this.runnable = runnable;
+    this.scheduler = scheduler;
+    this.precondition = precondition;
+    this.lock = lock;
+  }
+
+  /** Schedules a new one or reschedules an existing one. */
+  void reschedule(long delay, TimeUnit timeUnit) {
+    Future<?> existingTask = null;
+    FutureRunnable newWakeUp;
+
+    synchronized (lock) {
+      FutureRunnable existingWakeUp = wakeUp;
+      if (existingWakeUp != null) {
+        // cancel the existing one and schedule new one
+        existingWakeUp.cancelled = true;
+        existingTask = existingWakeUp.task;
+      } else if (precondition != null && !precondition.met()) {
```

Comment:
Is the point of the precondition just to avoid holding the lock when calling `schedule()`?